### PR TITLE
5457-provide dataset/version/filemetadata to External tools

### PR DESCRIPTION
**Pull request by qqmyers**
## New Contributors

Welcome! New contributors should at least glance at [CONTRIBUTING.md](/CONTRIBUTING.md), especially the section on pull requests where we encourage you to reach out to other developers before you start coding. Also, please note that we measure code coverage and prefer you write unit tests. Pull requests can still be reviewed without tests or completion of the checklist outlined below. Thanks!

## Related Issues

- connects to #5457 External tools can't see the dataset context

## Pull Request Checklist

- [ ] Unit [tests][] completed
- [ ] Integration [tests][]: None
- [ ] Deployment requirements, [SQL updates][], [Solr updates][], etc.: None
- [ ] [Documentation][docs] completed
- [ ] Merged latest from "develop" [branch][] and resolved conflicts

[tests]: http://guides.dataverse.org/en/latest/developers/testing.html
[SQL updates]: http://guides.dataverse.org/en/latest/developers/sql-upgrade-scripts.html
[Solr updates]: https://github.com/IQSS/dataverse/blob/develop/conf/solr/7.3.0/schema.xml
[docs]: http://guides.dataverse.org/en/latest/developers/documentation.html
[branch]: http://guides.dataverse.org/en/latest/developers/branching-strategy.html

### CODE REVIEW

File: src/main/java/edu/harvard/iq/dataverse/FileDownloadServiceBean.java [hunk 1]
```diff
@@ -235,6 +235,10 @@ public void explore(GuestbookResponse guestbookResponse, FileMetadata fmd, Exter
                 dataFile = guestbookResponse.getDataFile();
             }
         }
+        //For tools to get the dataset and datasetversion ids, we need a full DataFile object (not a findCheapAndEasy() copy)
```
Comment (scolapasta) [inline src/main/java/edu/harvard/iq/dataverse/FileDownloadServiceBean.java]: @qqmyers this looks fine but have you seen any performance issue here? If so, we could right some helper method like doesExternalToolNeedDataset(externalTool). But if there's trivial performance impact, not worth it.

Comment (qqmyers) [reply src/main/java/edu/harvard/iq/dataverse/FileDownloadServiceBean.java]: @scolapasta - I think it's trivial - this is only called during tool launch and it is then a lookup of one datafile object. I didn't see any obvious delay launching a tool and I think the main concerns previously have been for cases where many datafiles were being looked up, e.g. when scanning through the list in a dataset...

Comment (scolapasta) [reply src/main/java/edu/harvard/iq/dataverse/FileDownloadServiceBean.java]: OK< great. That is what I assumed as well, and why I had already moved to QA. :)

File: src/main/java/edu/harvard/iq/dataverse/externaltools/ExternalTool.java [hunk 2]
```diff
@@ -174,7 +174,9 @@ public JsonObjectBuilder toJson() {
         // from https://swagger.io/specification/#fixed-fields-29 but that's for URLs.
         FILE_ID("fileId"),
         SITE_URL("siteUrl"),
-        API_TOKEN("apiToken");
+        API_TOKEN("apiToken"),
+        DATASET_ID("datasetId"),
+        DATASET_VERSION("datasetVersion");
```
Comment (pdurbin) [inline src/main/java/edu/harvard/iq/dataverse/externaltools/ExternalTool.java]: @qqmyers can you please add these two new reserved words (datasetId and datasetVersion) to the list of possibilities? The current list can be found at http://guides.dataverse.org/en/4.10.1/installation/external-tools.html

**Pull request merged**